### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `IfStmtBinding` silently reorders a scalar read past an atomic write to the same buffer, returning the post-atomic value

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2594]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11 (built from source at cd37ed5f).

### System information

NVIDIA H100 80GB HBM3 (sm_90), Python 3.12, torch 2.12.1+cu130 (CUDA 13.0), tilelang 0.1.11 (tag `v0.1.11` = `cd37ed5f`). `IfStmtBinding` and the collector below are architecture-independent host-side TIR passes; the wrong result is not GPU-specific (reproduced on sm_90 this session).

### Problem description

Inside an `if COND:` (with **no** `else`), a scalar bind that reads a buffer just before an `atomic_*` to that **same** buffer is inlined *past* the atomic, so the value used downstream is the **post-atomic** value instead of the **pre-atomic** one. The kernel compiles and runs to completion with no error or warning; the result is silently wrong and deterministic.

Minimal trigger — a single thread, single element (no race is possible, so this is pure intra-thread statement reordering):

```python
if A[0] > 0:                 # if-WITHOUT-else
    v = A[0]                 # snapshot OLD A[0] (== 7)
    T.atomic_add(A[0], 100)  # A[0] += 100
    Out[0] = v               # MUST store the OLD A[0] == 7
```

With `A=[7]` the kernel emits `Out[0] = A[0];` (the bind `v` was substituted, and `A[0]` is now read *after* the atomic) and returns `Out[0] = 107`, not `7`. The wrong value affects the whole scalar atomic family (`atomic_add`/`atomic_max`/`atomic_min` and their `_ret` forms — all lower to `atomic_*_elem_op`); `atomic_add` and `atomic_max` both reproduced.

This is not a regression in the sense of "worked before" — the replayable-bind inlining path was **added** by #2245 (see Additional context).

<details>
<summary>Trigger boundary — two controls that isolate the cause (both correct)</summary>

Both controls run the identical read/write/read shape over the same inputs and are correct, so the wrong result is not a harness artifact:

- **Plain store instead of atomic** (`A[tx] = B[tx]`): the collector records the `BufferStore` as a write, so `v = A[tx]` is *not* classed replayable; the bind is kept (`int v = A[tx];` then `Out[tx] = v;`) → correct (OLD `A`).
- **Same atomic but with an `else` branch**: an `IfThenElse` that has an `else` never enters the inlining path (`if_stmt_binding.cc:245` returns before `:252`), so the bind is kept → correct (OLD `A`).

So the defect is exactly {atomic write invisible to the collector} × {`if`-without-`else` inline path}.
</details>

### Reproducible example code

```python
import tilelang
import tilelang.language as T
import torch

@tilelang.jit(out_idx=[1])
def bug():
    @T.prim_func
    def main(A: T.Tensor((1,), "int32"), Out: T.Tensor((1,), "int32")):
        with T.Kernel(1, threads=1):
            if A[0] > 0:                 # if without else -> inlining path
                v = A[0]                 # snapshot OLD A[0]
                T.atomic_add(A[0], 100)  # A[0] += 100
                Out[0] = v               # MUST be the OLD value
    return main

# CONTROL: same shape, atomic replaced by a plain store (collector sees the write)
@tilelang.jit(out_idx=[2])
def control():
    @T.prim_func
    def main(A: T.Tensor((1,), "int32"), B: T.Tensor((1,), "int32"), Out: T.Tensor((1,), "int32")):
        with T.Kernel(1, threads=1):
            if A[0] > 0:
                v = A[0]
                A[0] = B[0]
                Out[0] = v
    return main

kb = bug()
print("emitted:", [l.strip() for l in kb.get_kernel_source().splitlines() if "Out[0]" in l][0])
# -> "Out[0] = A[0];"   (the read was moved after the atomic)
out = kb(torch.tensor([7], dtype=torch.int32, device="cuda"))
print("bug   Out[0] =", int(out[0]))       # -> 107   (WRONG; expected 7)

kc = control()
out = kc(torch.tensor([7], dtype=torch.int32, device="cuda"),
         torch.tensor([999], dtype=torch.int32, device="cuda"))
print("ctrl  Out[0] =", int(out[0]))       # -> 7     (correct)
```

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic (reproduced twice, identical).
```

### Expected behavior

`Out` should hold the value of `A[0]` as read *before* the atomic (`7`), matching the source order of the three statements. When a bind reads a buffer that a later statement in the same guarded sequence mutates (including via an atomic), keeping the bind un-inlined across that mutation would preserve source order — the same way the plain-`BufferStore` case is already handled, and the same way #2278 already excludes pointer/handle binds from this inlining.

### Additional context

**Root cause.** `IfStmtBindingRewriter::IsReplayableBindStmt` calls `IsReplayableScalarBind`, which rejects a bind only if one of its read buffers is in `write_buffers` ([`bind_utils.h#L21-L41`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/common/bind_utils.h#L21-L41)). `write_buffers` is built by `IfStmtAccessCollector`, which records a write only from a `BufferStore` or a tile-op's `GetAccessRegions` ([`if_stmt_binding.cc#L42-L110`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/if_stmt_binding.cc#L42-L110)).

`T.atomic_add(A[tx], k)` lowers to `tl.atomic_add_elem_op(tl.access_ptr(A[tx], "rw"), k)` ([`atomic.py#L235-L245`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/language/atomic.py#L235-L245)) — a bare builtin (`atomic_add_elem_op`), not a tile-op, so `ParseOperator` returns undefined and the tile-op write branch is skipped. The collector then only recurses into the args, where it hits the `BufferLoad` of `A[tx]` inside `access_ptr` and calls `AddRead` ([`#L67-L70`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/if_stmt_binding.cc#L67-L70)). The two pointer branches (`address_of`, `tvm_access_ptr`) also only `AddRead` and never inspect the `rw`/write bit ([`#L94-L102`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/if_stmt_binding.cc#L94-L102)). So `A` is never added to `write_buffers`; `v = A[tx]` passes as replayable and is substituted into `Out[tx] = v`, which now reads `A[tx]` after the atomic. `IfStmtBinding` runs before `LowerAccessPtr` ([`pipeline.py#L104`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/tilelang/cuda/pipeline.py#L104)), so the destination is still the frontend `tl.access_ptr`. The `if`-without-`else` case is the only one that reaches `BindIfStmtWithReplayableBindInlining` ([`#L241-L253`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/if_stmt_binding.cc#L241-L253)).

**Suggested fix.** Have `IfStmtAccessCollector` record a **write** for an `atomic_*_elem_op` destination (or, more generally, for any `tl.access_ptr` / `tvm_access_ptr` whose access mask has the write bit set), so `write_buffers` is complete and `IsReplayableScalarBind` conservatively keeps the bind. Verified by the plain-store control above: when the collector sees the write, the bind is not inlined and the result is correct.

**Provenance.** The replayable-scalar-bind inlining and this collector were introduced by [#2245](https://github.com/tile-ai/tilelang/pull/2245) (merged 2026-05-26). Its immediate follow-up [#2278](https://github.com/tile-ai/tilelang/pull/2278) ("Keep pointer binds out of replayable scalar inlining") already recognized that this inlining needs to exclude unsafe binds — it excluded pointer/handle binds and relies on the read-vs-write-buffer check for value binds; the atomic-write hole is a gap in that same check, not a documented limitation. Still live on tip-of-main (fetched this session: the collector is unchanged, still `AddRead`-only, with no atomic/`rw`-bit handling).

**Related.** Distinct from #2538 (`MergeIfStmt` merging two same-condition `if`s where the first body invalidates the second's read): same "missing write-invalidation → stale read" family, but a different pass (`merge_if_stmt.cc`) and a different mechanism (merging two `if`s, not inlining a bind). I searched the open and closed tracker and found no existing report of a read reordered past an atomic in `IfStmtBinding`.
